Adds Arel::FactoryMethods#cast(node, type)

drnic opened the pull request:
For example:

```
  product_table = Product.arel_table
  product_table.cast(product_table[:position], "integer")
```

Produces SQL:

```
  CAST("products"."position" as integer)
```

### Motivation / Background

`CAST(field as type)` is a widely supported SQL function. This PR adds native arel support for this named function with a `cast(field, type)` helper.
### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

drnic committed:
Adds Arel::FactoryMethods#cast(node, type)

For example:
  product_table = Product.arel_table
  product_table.cast(product_table[:position], "integer")

Produces SQL:
  CAST(position as integer)

rafaelfranca commented on activerecord/test/cases/arel/factory_methods_test.rb:
```diff
@@ -41,6 +41,14 @@ def test_lower
       end
+
+      def test_cast
+        false_node = @factory.create_false
```
CI is borken in all database adapters that aren't sqlite because False isn't transformed to `0` in the other databases. Can you change the test to use always `0` instead of the false node?

drnic replied:
Ok no problem. Converted it into a table/field node object.

rafaelfranca commented on activerecord/CHANGELOG.md:
````diff
@@ -1,3 +1,20 @@
+*   Adds `Arel::FactoryMethods#cast(node, type)`
+
+    For example:
+
+    ```ruby
+      product_table = Product.arel_table
+      product_table.cast(product_table[:position], "integer")
+    ```
+
+    Produces SQL:
+
+    ```sql
+      CAST(position as integer)
+    ```
+
+    *Dr Nic Williams*
+
````
Arel isn't part of the public API of the framework, so changes to it should not be documented to the CHANGELOG that only list changes to the public API of the framework
```suggestion
```

drnic replied:
Ok removed.

drnic commented:
I'll investigate the new build failures and make a fix

rafaelfranca committed:
Test the node instead of the result of the SQL

The result of the SQL is not the same on all databases, so writing a
test that would pass in all cases would mean conditionals here.

Since we already testing if the SQL results of a As node is correct in
the visitor tests, we only need to check if the As node attributes are
correct here.

rafaelfranca commented:
This error is just a quote discrepancy between the adapters. I pushed a commit to this PR with the fix. Since we are already testing the SQL result of visiting a `AS` node in the visitor tests, we can just assert if the attributes of the `As` node is correct in this test avoiding to have to add conditionals depending on the database adapter.

drnic commented:
Gotcha. It seemed "right" to confirm the final SQL; but I was naive to assume all the adapters printed the same thing. Thanks for making time to fix and the explanations @rafaelfranca

rafaelfranca commented:
Thank you for implementing the feature
